Fix encoding of UTF-16 surrage pairs and decoding of escaped slashes

klaussfreire opened the pull request:
UCS4 escape sequences aren't valid JSON, so when Py_UNICODE_WIDE
is defined, UTF-16 surrogate pairs need to be produced.

Decoding retains the surrogate pairs as separate code points
instead of joining them, something that perhaps should be
improved, but isn't invalid (just inconvenient)

The escape code `\/` in the JSON spec means `/`, but it was not being
properly decoded (python's own escaping doesn't behave like this,
so one cannot rely on python's unescaping to handle it)

klaussfreire committed:
Fix decoding of escaped slashes

The escape code \/ in the JSON spec means /, but it was not being
properly decoded (python's own escaping doesn't behave like this,
so one cannot rely on python's unescaping to handle it)

klaussfreire committed:
Fix encoding of UTF-16 surrogate pairs

UCS4 escape sequences aren't valid JSON, so when Py_UNICODE_WIDE
is defined, UTF-16 surrogate pairs need to be produced.

Decoding retains the surrogate pairs as separate code points
instead of joining them, something that perhaps should be
improved, but isn't invalid (just inconvenient)

klaussfreire committed:
Remove trailing whitespace

danpascu commented:
Thanks. Please update the test script to accommodate for the changes and also add tests for the newly covered cases.

klaussfreire committed:
Adapt test, and fix encoding

When encoding surrogate pairs, 0x10000 has to be subtracted from
the code point before encoding the pair

klaussfreire commented:
Done. Sorry for the delay, I was on vacation.

The existing case already covered it, it was just that the reference output was not valid JSON. Good thing you made me check because there was a bug even.

Now the output matches the stdlib's json output.

danpascu commented:
Thanks. I'll take a look and integrate this early January after the holidays.

Our main repository is kept in darcs here: http://devel.ag-projects.com/repositories/python-cjson/
and is mirrored on github. Unfortunately this mirroring is only 1 way, so I cannot make changes on github, which means I can not apply pull requests directly. If you want the authorship of the patch to be preserved you can make a patch against the darcs repository, if not I'll apply your patch and mirror it back to github.

klaussfreire commented:
I've never worked with darcs, so I'm happy to leave that to you
